File: src/data/courses/dsa/string-hashing-problems/hashing/cpp-hashing-internals/code/java/CppHashingInternals.java
// C++ Hashing Internals — Java Equivalent Reference
// This lesson focuses on C++ std::unordered_map/unordered_set internals.
// Java equivalent: java.util.HashMap / java.util.HashSet

import java.util.*;

class CppHashingInternals {
    public static void main(String[] args) {
        // Java HashMap is the equivalent of C++ unordered_map
        HashMap<String, Integer> map = new HashMap<>();
        map.put("key", 42);
        System.out.println(map.get("key")); // 42

        // Key differences from C++:
        // 1. Java uses hashCode() + equals() (not operator== and std::hash)
        // 2. Java auto-treeifies buckets at size 8 (red-black tree)
        // 3. Default load factor: 0.75 (vs 1.0 in C++)
        // 4. Java doesn't allow custom hash in constructor (override hashCode)
    }
}

File: src/data/courses/dsa/string-hashing-problems/hashing/cpp-hashing-internals/code/python/cpp-hashing-internals.py
# C++ Hashing Internals — Python Equivalent Reference
# This lesson focuses on C++ std::unordered_map/unordered_set internals.
# Python equivalent: dict / set

class CppHashingInternals:
    """
    Python dict is the equivalent of C++ unordered_map.
    
    Key differences from C++:
    1. Python uses __hash__() + __eq__() (not operator== and std::hash)
    2. Python dicts use open addressing (not chaining like C++)
    3. Python dicts maintain insertion order (since 3.7)
    4. Load factor ~2/3 triggers resize (vs 1.0 in C++)
    """
    def demo(self):
        d = {"key": 42}
        print(d["key"])  # 42
        
        s = {1, 2, 3}  # set = unordered_set equivalent
        print(2 in s)   # True, O(1) average
